client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -270,9 +277,42 @@ public void wrongHost(WrongHost wrongHost) {
             state.failConnection(new ConnectionFailedException()); // TODO: Probably something else.
         }
 
+        /**
+         * Invariants for segment sealed:
+         *   - SegmentSealed callback and write will not run concurrently.
+         *   - During the execution of the call back no new writes will be executed.
+         * Once the segment Sealed callback is executed successfully
+         *  - there will be no new writes to this segment.
+         *  - any write to this segment will throw a SegmentSealedException.
+         *  - any thread waiting on state.getEmptyInflightFuture() will get SegmentSealedException.
+         * @param segmentIsSealed SegmentIsSealed WireCommand.
+         */
         @Override
         public void segmentIsSealed(SegmentIsSealed segmentIsSealed) {
-            state.failConnection(new SegmentSealedException());
+            log.trace("Received SegmentSealed {}", segmentIsSealed);
+            if (state.sealEncountered.compareAndSet(false, true)) {
+                CompletableFuture.<Void>supplyAsync(() -> {
+                    /*
+                      Invariants for segment sealed:
```

Comment:
This comment about the invariants is repeated in the javadoc block above. The comment is good, but I'm not sure we need it twice. :-)